Shape_detection/include/CGAL/Shape_detection/Region_growing/Polygon_mesh/Plane_face_region.h

```diff
@@ -0,0 +1,443 @@
+    const FT max_distance = parameters::choose_parameter(
+      parameters::get_parameter(np, internal_np::maximum_distance), FT(1));
+    CGAL_precondition(max_distance >= FT(0));
+    m_distance_threshold = max_distance;
+
+    const FT max_angle = parameters::choose_parameter(
+      parameters::get_parameter(np, internal_np::maximum_angle), FT(25));
+    CGAL_precondition(max_angle >= FT(0) && max_angle <= FT(90));
+
+    m_min_region_size = parameters::choose_parameter(
+      parameters::get_parameter(np, internal_np::minimum_region_size), 1);
+    CGAL_precondition(m_min_region_size > 0);
+
+    const FT default_cos_value = static_cast<FT>(std::cos(CGAL::to_double(
+      (max_angle * static_cast<FT>(CGAL_PI)) / FT(180))));
+    const FT cos_value = parameters::choose_parameter(
+      parameters::get_parameter(np, internal_np::cosine_of_maximum_angle), default_cos_value);
+    CGAL_precondition(cos_value >= FT(0) && cos_value <= FT(1));
+    m_cos_value_threshold = cos_value;
+  }
+
+  /// @}
+
+  /// \name Access
+  /// @{
+
+  /*!
+    \brief implements `RegionType::region_index_map()`.
+
+    This function creates an empty property map that maps each face to a `std::size_t`.
+  */
+  Region_index_map region_index_map() {
+    return get(CGAL::dynamic_face_property_t<std::size_t>(), m_pmesh);
+  }
+
+  /*!
+    \brief implements `RegionType::primitive()`.
+
+    This function provides the support plane of the seed face.
+
+    \return Primitive parameters that fits the region.
+
+    \pre is_valid_region(region)
+    \pre update(region)
+  */
+  Primitive primitive() const {
+    return m_plane;
+  }
+
+  /*!
+    \brief implements `RegionType::is_part_of_region()`.
+
+    This function determines if the face `query` is within
+    the `maximum_distance` from the corresponding plane and if the angle between
+    its normal and the plane's normal is below the `maximum_angle`. If both conditions
+    are satisfied, it returns `true`, otherwise `false`.
+
+    \param query item of the query face
+
+    \pre `query` is a valid face_descriptor (`boost::graph_traits<PolygonMesh>::%face_descriptor`) of `pmesh`
+  */
+  bool is_part_of_region(
+    const Item query,
+    const Region&) const
+  {
+    if (m_cos_value_threshold==1 || m_distance_threshold == 0)
+    {
+      halfedge_descriptor h = halfedge(query, m_pmesh);
+      for (vertex_descriptor v : vertices_around_face(h, m_pmesh))
+      {
+        if (!coplanar(m_p, m_q, m_r, get(m_vpm, v)))
+          return false;
+      }
+      return true;
+    }
+    else
+    {
+      // test on distance of points to the plane of the seed face
+      const FT squared_distance_threshold = m_distance_threshold * m_distance_threshold;
+      halfedge_descriptor h = halfedge(query, m_pmesh);
+      for (vertex_descriptor v : vertices_around_face(h, m_pmesh))
+      {
+        //TODO: that's a bit dummy that we retest points that are already in the region...
```

Comment:
@soesau any thought?

Reply:
If we keep the information inside the class, it will only be released when the object is released. A vpm would need clearing. An unordered_set could make sense and keep the memory as low as possible.

However, I don't know how big the effort wasted here is. A plane point distance check is very fast.